Fix JDK 17 InaccessibleObjectException with infinispan

vmuzikar opened the pull request:
Closes #15607

vmuzikar committed:
Fix JDK 17 InaccessibleObjectException with infinispan

vmuzikar commented on testsuite/integration-arquillian/tests/base/pom.xml:
```diff
@@ -352,7 +352,7 @@
                                         <shmSize>${docker.database.shmsize}</shmSize>
                                         <ports>
-                                            <port>${docker.database.port}</port>
+                                            <port>${docker.database.port}:${docker.database.port}</port>
```
AFAIK this is a non-breaking change that fixes clustering tests on macOS.

vmuzikar commented on quarkus/dist/src/main/content/bin/kc.sh:
```diff
@@ -90,6 +90,14 @@ else
 fi
 
+# See also https://github.com/wildfly/wildfly-core/blob/7e5624cf92ebe4b64a4793a8c0b2a340c0d6d363/core-feature-pack/common/src/main/resources/content/bin/common.sh#L57-L60
+if [ "x$JAVA_ADD_OPENS" = "x" ]; then
+   JAVA_OPTS="--add-opens=java.base/java.util=ALL-UNNAMED --add-opens=java.base/java.util.concurrent=ALL-UNNAMED"
+else
+   echo "JAVA_ADD_OPENS already set in environment; overriding default settings with values: $JAVA_ADD_OPENS"
+fi
+JAVA_OPTS="$JAVA_OPTS $JAVA_ADD_OPENS"
```
Using a separate env var to allow overriding it if needed. Did not want to use `JAVA_OPTS` directly as some folks are overriding it e.g. to tweak memory settings, so keeping it separate.

vmuzikar commented:
Test run with JDK 17: https://github.com/vmuzikar/keycloak/actions/runs/4072934233

vmuzikar commented:
FIPS failures seem to be unrelated. Can you please confirm @mposolda?

mposolda commented:
> FIPS failures seem to be unrelated. Can you please confirm @mposolda?

@vmuzikar I think you're right, but would still prefer to have them OK. So re-triggered the FIPS test.

vmuzikar commented:
FIPS failed again, so it might be related after all.

Seems to be failing with:
```
Failed to process record: org.bouncycastle.tls.TlsFatalAlert: handshake_failure(40); No selectable cipher suite
```
@mposolda Does it ring any bells? I'm a bit lost with FIPS, TBH. :)

mposolda commented:
@vmuzikar I guess that after your change the custom `java.security` file is not taken into consideration. It is attached here in the testsuite https://github.com/keycloak/keycloak/blob/main/testsuite/integration-arquillian/tests/base/src/main/java/org/keycloak/testsuite/arquillian/containers/AbstractQuarkusDeployableContainer.java#L353

If you want to test it, you can try to run command similar to this before/after your changes:
```
./kc.sh start -J-Djava.security.properties=$KEYCLOAK_SOURCES/testsuite/integration-arquillian/servers/auth-server/common/fips/kc.ava.security
```
and then check for example what is the content of some security property. Like for example `Security.getProperty("ssl.KeyManagerFactory.algorithm")` before/after your change. If it is value from that file (`PKIX`), it works, but if it is default from `$JAVA_HOME/conf/security/java.security` then it does not work.

vmuzikar committed:
Fix bug, add tests

vmuzikar commented on quarkus/dist/src/main/content/bin/kc.bat:
```diff
@@ -141,6 +149,11 @@ if not errorlevel == 1 (
     set IS_HELP_SHORT=true
 )
 
+if "%PRINT_ENV%" == "true" (
```
Did not want to print `JAVA_OPTS` by default not to polute log.

Since it's now used mainly for debugging/testing purposes, using `PRINT_ENV` env var for now to enable it. If we find this useful, we might consider e.g. a documented `--verbose` switch.

vmuzikar commented:
@mposolda FIPS failures were caused a bug I introduced in `kc.sh`. I realized we don't actually have any tests for the logic around `JAVA_OPTS` so I added some.

This is ready for another round of reviews.

CC @pedroigor

vmuzikar commented:
jFTR – a successful test run with JDK 17: https://github.com/vmuzikar/keycloak/actions/runs/4145705749

mposolda commented:
@vmuzikar Cool, Thanks!
